src/commando/core.py

```diff
@@ -1,2 +1,124 @@
+
+
+def _pprint(repo, message):
+    print(f"{repo}> {message}")
+
+
+def _run(cwd, results, command, **kwargs):
+    result = subprocess.run(  # pylint:disable=subprocess-run-check
+        command, cwd=cwd, stdout=PIPE, stderr=STDOUT, text=True, **kwargs
+    )
+    results.append(result)
+    result.check_returncode()
+    return result
+
+
+def _process_repo(
+    repo, command, commit_message, branch, pr_title, pr_body
+):  # pylint:disable=too-many-arguments
+
+    # Collect the results of all commands run for this repo.
+    results = []
+
+    pprint = partial(_pprint, repo)
+
+    with tempfile.TemporaryDirectory() as tmpdirname:
+        run = partial(_run, tmpdirname, results)
+
+        try:
+            pprint("Cloning")
+            run(
+                [
+                    "gh",
+                    "repo",
+                    "clone",
+                    repo,
+                    tmpdirname,
+                    "--",
+                    # Ensure that the origin remote is named 'origin', the
+                    # default name could be overridden by the git config.
+                    "--origin",
+                    "origin",
+                    # Try to speed it up by not cloning tags since we don't need them.
+                    # Note that I don't think we can speed this up even more by
+                    # using a shallow clone: gh-pr-upsert needs the commits and
+                    # history in order to compare branches.
+                    "--no-tags",
+                ]
+            )
+
+            pprint("Creating branch")
+            run(
+                [
+                    "git",
+                    "switch",
+                    # Don't automatically set upstream configuration if there's
+                    # a remote tracking branch with the same name.
+                    "--no-guess",
+                    # Don't set upstream tracking configuratin.
+                    "--no-track",
+                    # Create a new branch or reset an existing one to origin/main.
+                    "--force-create",
+                    branch,
+                    "origin/main",
+                ]
+            )
+
+            pprint("Running command")
+            pprint(command)
+            run(command, shell=True)
+
+            pprint("Committing any changes")
+            run(["git", "add", "--all"])
+
+            run(
+                [
+                    "git",
+                    "commit",
+                    # Don't crash if there are no changes to commit.
+                    # If an empty commit is created gh-pr-upsert will do the
+                    # right thing and not create an empty PR.
+                    "--allow-empty",
+                    "--message",
+                    commit_message,
+                ]
+            )
+
+            pprint("Sending PR")
+            try:
+                result = run(["gh-pr-upsert", "--title", pr_title, "--body", pr_body])
```

Comment:
It may seem a bit weird for Commando to be calling gh-pr-upsert (which is another Python package) as a subprocess. Commando could depend on gh-pr-upsert as a Python dependency and use it as a Python library.

But gh-pr-upsert isn't a library: it's just a command line tool with a command line interface, it doesn't have a Python library interface.

I think a CLI makes sense for gh-pr-upsert since it's potentially useful as a standalone CLI tool or to be called from bash scripts. It's also handy for testing gh-pr-upsert without having to open a Python shell.

The CLI is also the right place for gh-pr-upsert's defaults to live because argparse has support for default option values.

If we wanted to add a Python interface to gh-pr-upsert we'd have to somehow change it so that it has a Python API that applies the same defaults as the CLI and that also accepts it working directory as an option.